criu/files-reg.c

```diff
@@ -1380,12 +1388,318 @@ static bool should_check_size(int flags)
+		goto err;
+
+	memcpy(*build_id, (void *) note_header, size);
+
+	munmap(file_header, mapped_size);
+	return size;
+
+err:
+	munmap(file_header, mapped_size);
+	return -1;
+}
+
+/*
+ * Finds the build-id of the file by checking if the file is an ELF file
+ * and then calling either the 32-bit or the 64-bit function as necessary.
+ * Returns the number of bytes of the build-id if it could be
+ * obtained, else -1.
+ */
+static int get_build_id(const int fd, const struct stat *fd_status,
+				unsigned char **build_id)
+{
+	char *start_addr, buf[SELFMAG+1];
+	size_t mapped_size;
+
+	if (read(fd, buf, SELFMAG+1) != SELFMAG+1)
+		return -1;
+
+	/* 
+	 * The first 4 bytes contain a magic number identifying the file as an
+	 * ELF file. They should contain the characters ‘\x7f’, ‘E’, ‘L’, and
+	 * ‘F’, respectively. These characters are together defined as ELFMAG.
+	 */
+	if (strncmp(buf, ELFMAG, SELFMAG))
+		return -1;
+
+	/* 
+	 * If the build-id exists, then it will most likely be present in the
+	 * beginning of the file. Therefore at most only the first 1 MB of the 
+	 * file is mapped.
+	 */
+	mapped_size = min_t(size_t, fd_status->st_size, BUILD_ID_MAP_SIZE);
+	start_addr = (char *) mmap(0, mapped_size, PROT_READ, MAP_PRIVATE | MAP_FILE, fd, 0);
+	if (start_addr == MAP_FAILED) {
+		pr_warn("Couldn't mmap file with fd %d", fd);
+		return -1;
+	}
+
+	if (buf[EI_CLASS] == ELFCLASS32)
+		return get_build_id_32((Elf32_Ehdr *) start_addr, build_id, fd, mapped_size);
+	if (buf[EI_CLASS] == ELFCLASS64)
+		return get_build_id_64((Elf64_Ehdr *) start_addr, build_id, fd, mapped_size);
+	
+	munmap(start_addr, mapped_size);
+	return -1;
+}
+
+/*
+ * Finds and stores the build-id of a file, if it exists, so that it can be validated
+ * while restoring.
+ * Returns 1 if the build-id of the file could be stored, -1 if there was an error
+ * or 0 if the build-id could not be obtained.
+ */
+static int store_validation_data_build_id(RegFileEntry *rfe, int lfd,
+						const struct fd_parms *p)
+{
+	unsigned char *build_id = NULL;
+	int build_id_size, i;
+	int fd;
+
+	/*
+	 * Checks whether the file is atleast big enough to try and read the first
+	 * four (SELFMAG) bytes which should correspond to the ELF magic number
+	 * and the next byte which indicates whether the file is 32-bit or 64-bit.
+	 */
+	if (p->stat.st_size < SELFMAG+1)
+		return 0;
+
+	fd = open_proc(PROC_SELF, "fd/%d", lfd);
+	if (fd < 0) {
+		pr_err("Build-ID (For validation) could not be obtained for file %s because can't open the file\n",
+				rfe->name);
+		return -1;
+	}
+
+	build_id_size = get_build_id(fd, &(p->stat), &build_id);
+	close(fd);
+	if (!build_id || build_id_size == -1)
+		return 0;
+
+	rfe->build_id = xmalloc(sizeof(int) * build_id_size);
+	if (!rfe->build_id) {
+		pr_warn("Build-ID (For validation) could not be set for file %s\n",
+				rfe->name);
+		return -1;
+	}
+
+	rfe->n_build_id = build_id_size;
+	for (i = 0; i < build_id_size; i++)
+		rfe->build_id[i] = build_id[i];
```

Comment:
Bringing back the comment from the previous review:
> Build-id is a repeated uint32 in regfile.proto and build_id is an array of uchar, that's why I didn't use memcpy here

Allocate as:
`rfe->build_id = xzalloc(round_up(build_id_size, sizeof(uint32_t));`
That way you'll be able to use `memcpy()`.

Currently, by copying bytes into u32, you don't use 3/4 of the space in the array.

Reply:
@0x7f454c46 I'm not sure why, but this change seems to cause job 7 to fail (https://travis-ci.org/github/AnorexicAtticusFinch/criu/jobs/708363457) but I've pushed this code anyway.